### PR TITLE
Itemlink command and integration

### PR DESCRIPTION
**Pull request by SapitoSucio**
* **Server Mode**:   Both

* **Description of Pull Request**: 

This is a continuation of Cydh's work https://github.com/rathena/rathena/pull/6441

I added the changes to generate the itemlink on several commands that show info about items.

### @Iteminfo
iteminfo

### @mobinfo
mobinfo

### @idsearch
idsearch

### @autolootid
alootid

###

### CODE REVIEW

File: doc/script_commands.txt [hunk 1]
```diff
@@ -10866,6 +10866,27 @@ If <char id> is specified, the specified player is used rather than the attached
 
 ---------------------------------------
 
+*itemlink(<item_id>,<refine>,<card0>,<card1>,<card2>,<card3>,<enchantgrade>{,<RandomIDArray>,<RandomValueArray>,<RandomParamArray>});
+
+Generates an item link string for an item that can be used for npctalk, message,
+dispbottom, and broadcast commands. The result is a clickable-item name just
+like SHIFT+Click from a player's inventory/cart/equipment window. In the mes
+command is still usable but the item name is not clickable.
```
Comment (Akkarinage) [inline doc/script_commands.txt:10874]:
```suggestion
like SHIFT+Click from a player's inventory/cart/equipment window. This command can be
used with mes but the item name will not be clickable. You should use the normal client
tags for displaying item links in mes dialogues, if the client supports them. 

```

**Review by Akkarinage — CHANGES_REQUESTED** [review]

File: src/map/script.cpp [hunk 2]
```diff
@@ -26528,6 +26528,34 @@ BUILDIN_FUNC(item_enchant){
...
+
+	FETCH(3, item.refine);
+	FETCH(4, item.card[0]);
+	FETCH(5, item.card[1]);
+	FETCH(6, item.card[2]);
+	FETCH(7, item.card[3]);
+	FETCH(8, item.enchantgrade);
+
+#if PACKETVER >= 20150225
+	if ( script_hasdata(st,9) && script_getitem_randomoption(st, nullptr, &item, "itemlink", 9) == SCRIPT_CMD_FAILURE ) {
```
Comment (Atemo) [inline src/map/script.cpp:26549]: Return empty string (or end the script)

File: src/map/script.cpp [hunk 3]
```diff
@@ -26528,6 +26528,34 @@ BUILDIN_FUNC(item_enchant){
...
 
+/**
+* Generate item link string for client
+* itemlink(<item_id>,<refine>,<card0>,<card1>,<card2>,<card3>,<enchantgrade>{,<RandomIDArray>,<RandomValueArray>,<RandomParamArray>});
+* @author [Cydh]
+**/
+BUILDIN_FUNC(itemlink)
+{
+	struct item item = {};
+	item.nameid = script_getnum(st, 2);
```
Comment (Atemo) [inline src/map/script.cpp:26539]: Checks if the id exists in the db

Comment (SapitoSucio) [reply src/map/script.cpp:26539]: done 😊

File: src/map/script.cpp (hunk 2 above)
Comment (SapitoSucio) [reply src/map/script.cpp:26554]: im not sure what you mean, sorry. Could you point me in the right direction?

Comment (Atemo) [reply src/map/script.cpp:26554]: The npc script is still running when SCRIPT_CMD_FAILURE is returned. But the itemlink function is supposed to return a string.
So return an empty string via `script_pushconststr(st,"");`
or stop the script with `st->state = END;` (before SCRIPT_CMD_FAILURE)

Comment (SapitoSucio) [reply src/map/script.cpp:26555]: done 😊

File: src/map/atcommand.cpp [hunk 4]
```diff
@@ -8262,11 +8268,16 @@ ACMD_FUNC(iteminfo)
...
+
 	for (const auto &result : item_array) {
 		std::shared_ptr<item_data> item_data = result.second;
-
+		itemid = item_data->nameid;
+		tmp_item.nameid = itemid;
+		
 		sprintf(atcmd_output, msg_txt(sd,1277), // Item: '%s'/'%s'[%d] (%u) Type: %s | Extra Effect: %s
-			item_data->name.c_str(),item_data->ename.c_str(),item_data->slots,item_data->nameid,
+			createItemLink(tmp_item).c_str(),item_data->nameid,
```
Comment (Atemo) [inline src/map/atcommand.cpp:8280]: When I tested a few years ago, the item name was unreadable if the item didn't exist in the client. Do you know if there is a way to check if the item exists in the client serverside?

Comment (SapitoSucio) [reply src/map/atcommand.cpp:8280]: > When I tested a few years ago, the item name was unreadable if the item didn't exist in the client. Do you know if there is a way to check if the item exists in the client serverside?

Sorry, no

**Comment by Akkarinage** [comment]
I know this will be a bit more work, but it'd be nice if this was available via a config instead, maybe in battle/feature.conf or something?

**Comment by SapitoSucio** [comment]
> I know this will be a bit more work, but it'd be nice if this was available via a config instead, maybe in battle/feature.conf or something?

Sure can do!

**Comment by SapitoSucio** [comment]
Took me a while, first time doing this stuff 🐸👌

**Comment by ecdarreola** [comment]
Doesn't seem to work with 2020-03-04RagexeRE as shown below.
image

File: src/map/atcommand.cpp [hunk 5]
```diff
@@ -7770,28 +7791,55 @@ ACMD_FUNC(mobinfo)
...
+						sprintf(atcmd_output2, " ⁕ %s  %02.02f%%", id->ename.c_str(), (float)droprate / 100);
+					else
+						sprintf(atcmd_output2, " ⁕ %s %02.02f%%", createItemLink(tmp_item).c_str(), (float)droprate / 100);
+				}
+				strcat(atcmd_output, atcmd_output2);
+				if (++j % 3 == 0) {
+					clif_displaymessage(fd, atcmd_output);
 				strcpy(atcmd_output, " ");
-			}
+				} else {
```
Comment (ecdarreola) [inline src/map/atcommand.cpp:7824]: This caused the output to double.

**Comment by SapitoSucio** [comment]
> Doesn't seem to work with 2020-03-04RagexeRE as shown below. image

Yeah, apparently, it's broken on certain build dates

**Comment by iberryRO** [comment]
I Merge PR #6441 with PR #7291 and test using @iteminfo and it appears that the map Server Crash.

**Comment by SapitoSucio** [comment]
> I Merge PR #6441 with PR #7291 and test using @iteminfo and it appears that the map Server Crash.

Why merging both?

**Comment by iberryRO** [comment]
> > I Merge PR #6441 with PR #7291 and test using @iteminfo and it appears that the map Server Crash.
> 
> Why merging both?

Yesterday i tested it on PR#6441 it works fine but when i test it separately using PR#7291 it appears that the map-server crashes or it doesn't fit the other PR i have #7024

**Comment by iberryRO** [comment]
> > > I Merge PR #6441 with PR #7291 and test using @iteminfo and it appears that the map Server Crash.
> > 
> > 
> > Why merging both?
> 
> Yesterday i tested it on PR #6441 it works fine but when i test it separately using PR #7291 it appears that the map-server crashes or it doesn't fit the other PR i have #7024

ok i have started a new project and it works fine

**Comment by Atemo** [comment]
Please report only the issues about the PR, the PR is not your blog

**Comment by Amethyszxc** [comment]
getting Map crash when doing @ii, im using 2022 client.

File: conf/battle/feature.conf [hunk 6]
```diff
@@ -119,3 +119,11 @@ feature.barter: on
...
 // Requires: 2019-11-06RagexeRE or later
 feature.barter_extended: on
+
+
+// Itemlink System on Item/Messages Related Commands (Note 1)
+// Generates <ITEML> string for an item and can be used for npctalk, message,
+// dispbottom, and broadcast commands. The result is clickable-item name just
+// like from SHIFT+Click from player's inventory/cart/equipment window.
+// Requires: 2010-00-00RagexeRE or later
+feature.itemlink: on
```
Comment (secretdataz) [inline conf/battle/feature.conf:129]: Missing newline

File: src/common/utilities.cpp [hunk 7]
```diff
@@ -114,3 +114,33 @@ bool rathena::util::safe_multiplication( int64 a, int64 b, int64& result ){
...
+};
+
+std::string rathena::util::base62_encode( uint32 val ){
+	std::string result = "";
+	while (val != 0) {
+		result = base62_dictionary[(val % 62)] + result;
+		val /= 62;
+	}
+	return result;
+}
```
Comment (secretdataz) [inline src/common/utilities.cpp:146]: Missing newline

File: src/map/atcommand.cpp [hunk 8]
```diff
@@ -8320,10 +8375,12 @@ ACMD_FUNC(whodrops)
 		sprintf(atcmd_output, msg_txt(sd,269), MAX_SEARCH); // Displaying first %d matches
 		clif_displaymessage(fd, atcmd_output);
 	}
+	struct item tmp_item = {};
 	for (const auto &result : item_array) {
 		std::shared_ptr<item_data> id = result.second;
+		tmp_item.nameid = id->nameid;
 
-		sprintf(atcmd_output, msg_txt(sd,1285), id->ename.c_str(), id->slots, id->nameid); // Item: '%s'[%d] (ID:%u)
+		sprintf(atcmd_output, msg_txt(sd,1285), createItemLink(tmp_item).c_str(), id->nameid); // Item: %s (ID: %u)
```
Comment (secretdataz) [inline src/map/atcommand.cpp:8383]: battle config check here

File: src/map/atcommand.cpp [hunk 9]
```diff
@@ -9319,9 +9377,9 @@ ACMD_FUNC(itemlist)
 		}
 
 		if( it->refine )
-			StringBuf_Printf(&buf, "%d %s %+d (%s, id: %u)", it->amount, itd->ename.c_str(), it->refine, itd->name.c_str(), it->nameid);
+			StringBuf_Printf(&buf, "⁕ %d %s %+d (%s, id: %u)", it->amount, createItemLink(tmp_item).c_str(), it->refine, itd->name.c_str(), it->nameid);
```
Comment (secretdataz) [inline src/map/atcommand.cpp:9380]: battle config check here

File: src/map/atcommand.cpp [hunk 10]
```diff
@@ -9319,9 +9377,9 @@ ACMD_FUNC(itemlist)
...
 		else
-			StringBuf_Printf(&buf, "%d %s (%s, id: %u)", it->amount, itd->ename.c_str(), itd->name.c_str(), it->nameid);
+			StringBuf_Printf(&buf, "⁕ %d %s (%s, id: %u)", it->amount, createItemLink(tmp_item).c_str(), itd->name.c_str(), it->nameid);
```
Comment (secretdataz) [inline src/map/atcommand.cpp:9382]: battle config check here

File: src/map/atcommand.cpp [hunk 11]
```diff
@@ -9424,7 +9482,7 @@ ACMD_FUNC(itemlist)
 				if( counter2 != 1 )
 					StringBuf_AppendStr(&buf, ", ");
 
-				StringBuf_Printf(&buf, "#%d %s (id: %u)", counter2, card->ename.c_str(), card->nameid);
+				StringBuf_Printf(&buf, "#%d %s (id: %u)", counter2, createItemLink(tmp_item).c_str(), card->nameid);
```
Comment (secretdataz) [inline src/map/atcommand.cpp:9485]: battle config check here

**Review by secretdataz — CHANGES_REQUESTED** [review]
Please add back the test cases from the original PR (in the npc directory).

**Review by secretdataz — COMMENTED** [review]
Looks good now. Doesn't compile tho.

**Review by secretdataz — APPROVED** [review]

File: conf/msg_conf/map_msg.conf [hunk 12]
```diff
@@ -1793,5 +1793,10 @@
...
 1520: %d items have been graded.
 
+// ItemLink
+1521: Item: %s (ID: %u)
+1522: Item: %s ⁕ (ID: %u) Type: %s | Extra Effect: %s
+1523:⠀⁕ %s : %u
+
 //Custom translations
-import: conf/msg_conf/import/map_msg_eng_conf.txt
+import: conf/msg_conf/import/map_msg_eng_conf.txt
```
Comment (Atemo) [inline conf/msg_conf/map_msg.conf:1802]: Missing newline

File: npc/test/ci/0000_funcs.txt [hunk 13]
```diff
@@ -0,0 +1,14 @@
+function	script	AssertTrue	{
+	if (!getarg(0)) {
+		errormes "AssertTrue failed for " + getarg(1) + ".";
+		return false;
+	}
```
Comment (Atemo) [inline npc/test/ci/0000_funcs.txt:5]:
```suggestion
	}
	return true;
```

File: src/map/atcommand.cpp [hunk 14]
```diff
@@ -7770,28 +7791,36 @@ ACMD_FUNC(mobinfo)
...
+					if (!battle_config.feature_itemlink)
+						sprintf(atcmd_output2, " ⁕ %s  %02.02f%%", id->ename.c_str(), (float)droprate / 100);
+					else
+						sprintf(atcmd_output2, " ⁕ %s %02.02f%%", createItemLink(tmp_item).c_str(), (float)droprate / 100);
+				}
+				strcat(atcmd_output, atcmd_output2);
+				if (++j % 3 == 0) {
+					clif_displaymessage(fd, atcmd_output);
+					strcpy(atcmd_output, " ");
+				}
```
Comment (Atemo) [inline src/map/atcommand.cpp:7823]: It looks like there is an extra indent

File: src/map/atcommand.cpp [hunk 15]
```diff
@@ -7811,28 +7840,42 @@ ACMD_FUNC(mobinfo)
...
+										sprintf(atcmd_output2, " %s[%d] %02.02f%%", id->ename.c_str(), id->slots, mvppercent);
+									else
+										sprintf(atcmd_output2, " %s %02.02f%%", createItemLink(tmp_item).c_str(), mvppercent);							
+								} else {
+										if (!battle_config.feature_itemlink)
+											sprintf(atcmd_output2, " ⁕ %s %02.02f%%", id->ename.c_str(), mvppercent);
+										else
+											sprintf(atcmd_output2, " ⁕ %s %02.02f%%", createItemLink(tmp_item).c_str(), mvppercent);
+								}
+						}
```
Comment (Atemo) [inline src/map/atcommand.cpp:7878]: It looks like there is an extra indent

File: src/map/atcommand.cpp [hunk 16]
```diff
@@ -8262,14 +8305,25 @@ ACMD_FUNC(iteminfo)
...
 	}
+
+	struct item tmp_item = {};
+
 	for (const auto &result : item_array) {
 		std::shared_ptr<item_data> item_data = result.second;
-
-		sprintf(atcmd_output, msg_txt(sd,1277), // Item: '%s'/'%s'[%d] (%u) Type: %s | Extra Effect: %s
+		itemid = item_data->nameid;
+		tmp_item.nameid = itemid;
```
Comment (Atemo) [inline src/map/atcommand.cpp:8314]: Can be in the else block (below)

File: src/map/atcommand.cpp [hunk 17]
```diff
@@ -8262,14 +8305,25 @@ ACMD_FUNC(iteminfo)
...
+		if (!battle_config.feature_itemlink)
+			sprintf(atcmd_output, msg_txt(sd,1277), // Item: %s ⁕ (ID: %u) Type: %s | Extra Effect: %s
 			item_data->name.c_str(),item_data->ename.c_str(),item_data->slots,item_data->nameid,
 			(item_data->type != IT_AMMO) ? itemdb_typename((enum item_types)item_data->type) : itemdb_typename_ammo((e_ammo_type)item_data->subtype),
-			(item_data->script==NULL)? msg_txt(sd,1278) : msg_txt(sd,1279) // None / With script
-		);
+			(item_data->script==NULL)? msg_txt(sd,1278) : msg_txt(sd,1279)); // None / With script
+
+		else
```
Comment (Atemo) [inline src/map/atcommand.cpp:8321]: Missing bracket

File: src/map/script.cpp [hunk 18]
```diff
@@ -26712,6 +26712,40 @@ BUILDIN_FUNC(item_enchant){
...
+
+	FETCH(3, item.refine);
+	FETCH(4, item.card[0]);
+	FETCH(5, item.card[1]);
+	FETCH(6, item.card[2]);
+	FETCH(7, item.card[3]);
+	FETCH(8, item.enchantgrade);
+
+#if PACKETVER >= 20150225
+	if ( script_hasdata(st,9) && script_getitem_randomoption(st, nullptr, &item, "itemlink", 9) == SCRIPT_CMD_FAILURE ) {
```
Comment (Atemo) [inline src/map/script.cpp:26739]:
```suggestion
	if ( script_hasdata(st,9) && script_getitem_randomoption(st, nullptr, &item, "itemlink", 9) == false) {
```

File: src/map/atcommand.cpp [hunk 19]
```diff
@@ -7823,15 +7852,29 @@ ACMD_FUNC(mobinfo)
...
+								if (!battle_config.feature_itemlink)
+									sprintf(atcmd_output2, " %s[%d] %02.02f%%", id->ename.c_str(), id->slots, mvppercent);
+								else
+									sprintf(atcmd_output2, " %s %02.02f%%", createItemLink(tmp_item).c_str(), mvppercent);							
+							} else {
+									if (!battle_config.feature_itemlink)
+										sprintf(atcmd_output2, " ⁕ %s %02.02f%%", id->ename.c_str(), mvppercent);
+									else
+										sprintf(atcmd_output2, " ⁕ %s %02.02f%%", createItemLink(tmp_item).c_str(), mvppercent);
+							}
```
Comment (Atemo) [inline src/map/atcommand.cpp:7877]: (just another extra indentation)
```suggestion
						if (id->slots){
							if (!battle_config.feature_itemlink)
								sprintf(atcmd_output2, " %s[%d] %02.02f%%", id->ename.c_str(), id->slots, mvppercent);
							else
								sprintf(atcmd_output2, " %s %02.02f%%", createItemLink(tmp_item).c_str(), mvppercent);							
						} else {
							if (!battle_config.feature_itemlink)
								sprintf(atcmd_output2, " ⁕ %s %02.02f%%", id->ename.c_str(), mvppercent);
							else
								sprintf(atcmd_output2, " ⁕ %s %02.02f%%", createItemLink(tmp_item).c_str(), mvppercent);
						}
```

**Review by Atemo — APPROVED** [review]

File: src/map/itemdb.cpp [hunk 20]
```diff
@@ -4529,6 +4531,86 @@ void itemdb_reload(void) {
...
+// All these dates are unconfirmed
+#if PACKETVER_MAIN_NUM >= 20200916 || PACKETVER_RE_NUM >= 20200724
+	const std::string start_tag = "<ITEML>";
+	const std::string closing_tag = "</ITEML>";
+#elif PACKETVER >= 20151104
+	const std::string start_tag = "<ITEM>";
+	const std::string closing_tag = "</ITEM>";
+#elif PACKETVER >= 20100000
+	const std::string start_tag = "<ITEMLINK>";
+	const std::string closing_tag = "</ITEMLINK>";
```
Comment (Lemongrass3110) [inline src/map/itemdb.cpp:4559]: I partially broke stuff here. I mixed two different features, but dont worry, we will clean it up.

**Review by Lemongrass3110 — COMMENTED** [review]
Applies to the whole PR:
Instead of checking the config on each reference, why not check it inside createItemLink itself so you do not need to add if/else logic to each call?

Also it seems like you or whoever likes * more than -
Are those changes really necessary?

File: src/map/atcommand.cpp [hunk 21]
```diff
@@ -7770,28 +7791,36 @@ ACMD_FUNC(mobinfo)
 			drop_modifier = pc_level_penalty_mod( sd, PENALTY_DROP, mob );
 		}
 #endif
-
+		struct item tmp_item = {};
```
Comment (aleos89) [inline src/map/atcommand.cpp:7794]:
```suggestion
		struct item tmp_item = {};

```

File: src/map/atcommand.cpp [hunk 22]
```diff
@@ -7823,15 +7852,29 @@ ACMD_FUNC(mobinfo)
...
 					if (j == 1) {
-						if (id->slots)
-							sprintf(atcmd_output2, " %s[%d]  %02.02f%%", id->ename.c_str(), id->slots, mvppercent);
-						else
-							sprintf(atcmd_output2, " %s  %02.02f%%", id->ename.c_str(), mvppercent);
+						if (id->slots){
+							if (!battle_config.feature_itemlink)
+								sprintf(atcmd_output2, " %s[%d] %02.02f%%", id->ename.c_str(), id->slots, mvppercent);
+							else
+								sprintf(atcmd_output2, " %s %02.02f%%", createItemLink(tmp_item).c_str(), mvppercent);
```
Comment (aleos89) [inline src/map/atcommand.cpp:7859]:
```suggestion
								sprintf(atcmd_output2, " %s %02.02f%%", itemdb_createItemLink(tmp_item).c_str(), mvppercent);
```

File: src/map/atcommand.cpp [hunk 23]
```diff
@@ -7823,15 +7852,29 @@ ACMD_FUNC(mobinfo)
...
+						} else {
+								if (!battle_config.feature_itemlink)
+									sprintf(atcmd_output2, " ⁕ %s %02.02f%%", id->ename.c_str(), mvppercent);
+								else
+									sprintf(atcmd_output2, " ⁕ %s %02.02f%%", createItemLink(tmp_item).c_str(), mvppercent);
```
Comment (aleos89) [inline src/map/atcommand.cpp:7864]:
```suggestion
									sprintf(atcmd_output2, " ⁕ %s %02.02f%%", itemdb_createItemLink(tmp_item).c_str(), mvppercent);
```

File: src/map/atcommand.cpp [hunk 24]
```diff
@@ -7823,15 +7852,29 @@ ACMD_FUNC(mobinfo)
...
 					} else {
-						if (id->slots)
-							sprintf(atcmd_output2, " - %s[%d]  %02.02f%%", id->ename.c_str(), id->slots, mvppercent);
-						else
-							sprintf(atcmd_output2, " - %s  %02.02f%%", id->ename.c_str(), mvppercent);
+						if (id->slots){
+							if (!battle_config.feature_itemlink)
+								sprintf(atcmd_output2, " %s[%d] %02.02f%%", id->ename.c_str(), id->slots, mvppercent);
+							else
+								sprintf(atcmd_output2, " %s %02.02f%%", createItemLink(tmp_item).c_str(), mvppercent);
```
Comment (aleos89) [inline src/map/atcommand.cpp:7871]:
```suggestion
								sprintf(atcmd_output2, " %s %02.02f%%", itemdb_createItemLink(tmp_item).c_str(), mvppercent);							
```

File: src/map/atcommand.cpp [hunk 25]
```diff
@@ -7823,15 +7852,29 @@ ACMD_FUNC(mobinfo)
...
+						if (id->slots){
+							if (!battle_config.feature_itemlink)
+								sprintf(atcmd_output2, " %s[%d] %02.02f%%", id->ename.c_str(), id->slots, mvppercent);
+							else
+								sprintf(atcmd_output2, " %s %02.02f%%", createItemLink(tmp_item).c_str(), mvppercent);							
+						} else {
+							if (!battle_config.feature_itemlink)
+								sprintf(atcmd_output2, " ⁕ %s %02.02f%%", id->ename.c_str(), mvppercent);
+							else
+								sprintf(atcmd_output2, " ⁕ %s %02.02f%%", createItemLink(tmp_item).c_str(), mvppercent);
```
Comment (aleos89) [inline src/map/atcommand.cpp:7876]:
```suggestion
								sprintf(atcmd_output2, " ⁕ %s %02.02f%%", itemdb_createItemLink(tmp_item).c_str(), mvppercent);
```

File: src/map/atcommand.cpp [hunk 26]
```diff
@@ -8320,10 +8375,15 @@ ACMD_FUNC(whodrops)
 		sprintf(atcmd_output, msg_txt(sd,269), MAX_SEARCH); // Displaying first %d matches
 		clif_displaymessage(fd, atcmd_output);
 	}
+	struct item tmp_item = {};
```
Comment (aleos89) [inline src/map/atcommand.cpp:8378]:
```suggestion
```

File: src/map/atcommand.cpp [hunk 27]
```diff
@@ -8320,10 +8375,15 @@ ACMD_FUNC(whodrops)
...
 	for (const auto &result : item_array) {
 		std::shared_ptr<item_data> id = result.second;
+		tmp_item.nameid = id->nameid;
```
Comment (aleos89) [inline src/map/atcommand.cpp:8381]:
```suggestion
		struct item tmp_item = {};

		tmp_item.nameid = id->nameid;
```

File: src/map/atcommand.cpp [hunk 28]
```diff
@@ -8320,10 +8375,15 @@ ACMD_FUNC(whodrops)
...
 
-		sprintf(atcmd_output, msg_txt(sd,1285), id->ename.c_str(), id->slots, id->nameid); // Item: '%s'[%d] (ID:%u)
+		if (battle_config.feature_itemlink)
+			sprintf(atcmd_output, msg_txt(sd,1521), createItemLink(tmp_item).c_str(), id->nameid); // Item: %s (ID: %u)
```
Comment (aleos89) [inline src/map/atcommand.cpp:8384]:
```suggestion
			sprintf(atcmd_output, msg_txt(sd,1521), itemdb_createItemLink(tmp_item).c_str(), id->nameid); // Item: %s (ID: %u)
```

File: src/map/atcommand.cpp [hunk 29]
```diff
@@ -9319,9 +9380,15 @@ ACMD_FUNC(itemlist)
 		}
 
 		if( it->refine )
-			StringBuf_Printf(&buf, "%d %s %+d (%s, id: %u)", it->amount, itd->ename.c_str(), it->refine, itd->name.c_str(), it->nameid);
+			if (battle_config.feature_itemlink)
+				StringBuf_Printf(&buf, "⁕ %d %s %+d (%s, id: %u)", it->amount, createItemLink(tmp_item).c_str(), it->refine, itd->name.c_str(), it->nameid);
```
Comment (aleos89) [inline src/map/atcommand.cpp:9384]:
```suggestion
				StringBuf_Printf(&buf, "⁕ %d %s %+d (%s, id: %u)", it->amount, itemdb_createItemLink(tmp_item).c_str(), it->refine, itd->name.c_str(), it->nameid);
```

File: src/map/atcommand.cpp [hunk 30]
```diff
@@ -9319,9 +9380,15 @@ ACMD_FUNC(itemlist)
...
+			else
+				StringBuf_Printf(&buf, "%d %s %+d (%s, id: %u)", it->amount, itd->ename.c_str(), it->refine, itd->name.c_str(), it->nameid);
 		else
-			StringBuf_Printf(&buf, "%d %s (%s, id: %u)", it->amount, itd->ename.c_str(), itd->name.c_str(), it->nameid);
+			if (battle_config.feature_itemlink)
+				StringBuf_Printf(&buf, "⁕ %d %s (%s, id: %u)", it->amount, createItemLink(tmp_item).c_str(), itd->name.c_str(), it->nameid);
```
Comment (aleos89) [inline src/map/atcommand.cpp:9389]:
```suggestion
				StringBuf_Printf(&buf, "⁕ %d %s (%s, id: %u)", it->amount, itemdb_createItemLink(tmp_item).c_str(), itd->name.c_str(), it->nameid);
```

File: src/map/atcommand.cpp [hunk 31]
```diff
@@ -9424,7 +9491,10 @@ ACMD_FUNC(itemlist)
 				if( counter2 != 1 )
 					StringBuf_AppendStr(&buf, ", ");
 
-				StringBuf_Printf(&buf, "#%d %s (id: %u)", counter2, card->ename.c_str(), card->nameid);
+				if (battle_config.feature_itemlink)
+					StringBuf_Printf(&buf, "#%d %s (id: %u)", counter2, createItemLink(tmp_item).c_str(), card->nameid);
```
Comment (aleos89) [inline src/map/atcommand.cpp:9495]:
```suggestion
					StringBuf_Printf(&buf, "#%d %s (id: %u)", counter2, itemdb_createItemLink(tmp_item).c_str(), card->nameid);
```

File: src/map/itemdb.hpp [hunk 32]
```diff
@@ -1612,6 +1613,8 @@ bool itemdb_parse_roulette_db(void);
 
 void itemdb_reload(void);
 
+std::string createItemLink(struct item& data);
```
Comment (aleos89) [inline src/map/itemdb.hpp:1616]:
```suggestion
std::string itemdb_createItemLink(struct item& data);
```

File: src/map/script.cpp [hunk 33]
```diff
@@ -26712,6 +26712,40 @@ BUILDIN_FUNC(item_enchant){
...
 }
 
+/**
+* Generate item link string for client
+* itemlink(<item_id>,<refine>,<card0>,<card1>,<card2>,<card3>,<enchantgrade>{,<RandomIDArray>,<RandomValueArray>,<RandomParamArray>});
+* @author [Cydh]
+**/
+BUILDIN_FUNC(itemlink)
+{
+	struct item item = {};
```
Comment (aleos89) [inline src/map/script.cpp:26722]:
```suggestion
	struct item item = {};

```

File: src/map/script.cpp [hunk 34]
```diff
@@ -26712,6 +26712,40 @@ BUILDIN_FUNC(item_enchant){
...
+	item.nameid = script_getnum(st, 2);
+	
+	if( !item_db.exists( item.nameid ) ){
+		ShowError( "Itemlink: Item ID %u does not exists.\n", item.nameid );
```
Comment (aleos89) [inline src/map/script.cpp:26726]:
```suggestion
		ShowError( "buildin_itemlink: Item ID %u does not exists.\n", item.nameid );
```

File: src/map/script.cpp [hunk 35]
```diff
@@ -26712,6 +26712,40 @@ BUILDIN_FUNC(item_enchant){
...
+	FETCH(8, item.enchantgrade);
+
+#if PACKETVER >= 20150225
+	if ( script_hasdata(st,9) && script_getitem_randomoption(st, nullptr, &item, "itemlink", 9) == false) {
+		st->state = END;
+		return SCRIPT_CMD_FAILURE;
+	}
+#endif
+
+	std::string itemlstr = createItemLink(item);
```
Comment (aleos89) [inline src/map/script.cpp:26745]:
```suggestion
	std::string itemlstr = itemdb_createItemLink(item);
```

File: src/map/atcommand.cpp [hunk 36]
```diff
@@ -9319,9 +9380,15 @@ ACMD_FUNC(itemlist)
...
+			if (battle_config.feature_itemlink)
+				StringBuf_Printf(&buf, "⁕ %d %s %+d (%s, id: %u)", it->amount, createItemLink(tmp_item).c_str(), it->refine, itd->name.c_str(), it->nameid);
+			else
+				StringBuf_Printf(&buf, "%d %s %+d (%s, id: %u)", it->amount, itd->ename.c_str(), it->refine, itd->name.c_str(), it->nameid);
 		else
-			StringBuf_Printf(&buf, "%d %s (%s, id: %u)", it->amount, itd->ename.c_str(), itd->name.c_str(), it->nameid);
+			if (battle_config.feature_itemlink)
+				StringBuf_Printf(&buf, "⁕ %d %s (%s, id: %u)", it->amount, createItemLink(tmp_item).c_str(), itd->name.c_str(), it->nameid);
+			else
+				StringBuf_Printf(&buf, "%d %s (%s, id: %u)", it->amount, itd->ename.c_str(), itd->name.c_str(), it->nameid);
```
Comment (aleos89) [inline src/map/atcommand.cpp:9391]:
```suggestion
				StringBuf_Printf(&buf, "%d %s (%s, id: %u)", it->amount, itd->ename.c_str(), itd->name.c_str(), it->nameid);
		}
```

File: src/map/atcommand.cpp [hunk 37]
```diff
@@ -9298,17 +9358,18 @@ ACMD_FUNC(itemlist)
 
 	count = 0; // total slots occupied
 	counter = 0; // total items found
+	struct item tmp_item = {};
```
Comment (aleos89) [inline src/map/atcommand.cpp:9361]:
```suggestion

	struct item tmp_item = {};

```

**Review by aleos89 — COMMENTED** [review]
Just a few changes/fixes!

**Review by Lemongrass3110 — APPROVED** [review]
seal

**Review by aleos89 — APPROVED** [review]

**Review by secretdataz — APPROVED** [review]

**Review by Atemo — APPROVED** [review]

**Comment by jovvybersamin** [comment]
I'm trying to add this manually since my emulator is not in the latest version, I want to use this in my server.

However, I'm getting this error upon compiling.

image

**Pull request merged**